src/Database/Franz/Network.hs

```diff
@@ -171,72 +181,95 @@ getResponse = do
+  -- fetch request. If anything goes wrong in the middle, we're
+  -- certain it'll get removed.
+  withSharedResource connStates reqId (newTVarIO WaitingInstant) (\_ -> pure ()) $ \reqVar -> do
+
+    let allowDelayedResponse = isRight onDelayed'e
+    -- Send the user request.
+    withMVar connSocket $ \sock -> SB.sendAll sock $ encode
+      $ RawRequest reqId req allowDelayedResponse
+    let -- Cancel the request with the server if we catch an
+        -- exception, such as user giving up on waiting and killing
+        -- the action.
+      cancelRequest = withMVar connSocket $ \sock ->
+        SB.sendAll sock $ encode $ RawClean reqId
+
+      getInstant = readTVar reqVar >>= \case
+        Errored e -> throwSTM e
+        WaitingInstant -> retry -- wait for an instant response
+        Available xs -> return $ Just xs
+        WaitingDelayed -> pure Nothing
+
+      getDelayed = readTVar reqVar >>= \case
+        WaitingDelayed -> retry
+        Available xs -> return xs
+        Errored e -> throwSTM e
+        WaitingInstant -> throwSTM $ ClientError $
+          "fetch/WaitingDelayed: unexpected state WaitingInstant"
+
+    instant'm <- atomically getInstant `onException` cancelRequest
+    case instant'm of
+      -- We got an answer instantly, run user-callback.
+      Just xs -> onInstant xs
+      -- We were asked to wait.
+      Nothing -> case onDelayed'e of
+        -- User is dis-interested in waiting for results. We have
+        -- indicated this in the request itself, therefore we don't
+        -- have to cancel anything and we can just run the default
+        -- user action.
+        Left dontWait -> dontWait
+        -- User is interested in waiting, at least for now. Initialise
+        -- any context the user may need and then wait for results.
+        Right onDelayedOuter -> do
+          -- Get result handler. If this action fails, we want to
+          -- cancel the request as we obviously won't be able to
+          -- handle any results anyway.
+          onDelayedResult <- onDelayedOuter `onException` cancelRequest
+          -- Actually wait for the result. If we fail (such as by user
+          -- killing the fetch action via a timeout), cancel the
+          -- request as we're no longer able to handle it even if it
+          -- comes in.
+          delayedResult <- atomically getDelayed `onException` cancelRequest
+          -- We got the result, we can finally handle it.
+          onDelayedResult delayedResult
 
 -- | Send a single query and wait for the result. If it timeouts, it returns an empty list.
 fetchSimple :: Connection
   -> Int -- ^ timeout in microseconds
   -> Query
   -> IO Contents
-fetchSimple conn timeout req = fetch conn req (fmap (maybe mempty id) . atomicallyWithin timeout . awaitResponse)
-
-atomicallyWithin :: Int -- ^ timeout in microseconds
-  -> STM a
-  -> IO (Maybe a)
-atomicallyWithin timeout m = do
-  d <- newDelay timeout
```

Comment:
Using `stm-delay` instead of forking a thread doing `threadDelay` is deliberate; In modern GHC, it uses GHC.Event internally.